agent/engine/serviceconnect/manager_linux.go

```diff
@@ -372,17 +370,40 @@ func (m *manager) AugmentInstanceContainer(task *apitask.Task, container *apicon
 	return nil
 }
 
-// LoadImage helps load the AppNetAgent container image for the agent
+func (agent *manager) setLoadedAppnetVerion(appnetInterfaceVersion string) {
+	agent.appnetInterfaceVersion = appnetInterfaceVersion
+}
+
+// LoadImage helps load the AppNetAgent container image for the agent latest supported
+// AppNet interface version by looking for the AppNet agent tar name from supported list
+// of AppNet versions from highest to lowest version when loading AppNet image
 func (agent *manager) LoadImage(ctx context.Context, _ *config.Config, dockerClient dockerapi.DockerClient) (*types.ImageInspect, error) {
-	logger.Debug("Loading appnet agent container tarball:", logger.Fields{
-		field.Image: agent.AgentContainerTarballPath,
-	})
-	if err := loader.LoadFromFile(ctx, agent.AgentContainerTarballPath, dockerClient); err != nil {
-		return nil, err
+	var loadErr error
+	for _, supportedAppnetInterfaceVersion := range getSupportedAppnetInterfaceVersions() {
+		agentContainerTarballPath := fmt.Sprintf(defaultAgentContainerTarballPathFormat, supportedAppnetInterfaceVersion)
+		if _, err := os.Stat(agentContainerTarballPath); err != nil {
+			logger.Warn(fmt.Sprintf("AppNet agent container tarball unavailable: %s", agentContainerTarballPath), logger.Fields{
+				field.Error: err,
+			})
+			continue
+		}
+		logger.Debug(fmt.Sprintf("Loading appnet agent container tarball: %s", agentContainerTarballPath))
+		if loadErr = loader.LoadFromFile(ctx, agentContainerTarballPath, dockerClient); loadErr != nil {
+			logger.Warn(fmt.Sprintf("Unable to load appnet agent container tarball: %s", agentContainerTarballPath),
+				logger.Fields{
+					field.Error: loadErr,
+				})
+			continue
+		}
+		agent.setLoadedAppnetVerion(supportedAppnetInterfaceVersion)
+		imageName, _ := agent.GetLoadedImageName()
+		logger.Info(fmt.Sprintf("Successfully loaded appnet agent container tarball: %s", agentContainerTarballPath),
+			logger.Fields{
+				field.Image: imageName,
+			})
+		return loader.GetContainerImage(imageName, dockerClient)
 	}
-
-	imageName, _ := agent.GetLoadedImageName()
-	return loader.GetContainerImage(imageName, dockerClient)
+	return nil, loadErr
```

Comment:
Not blocking and more a question. In addition to show the warning log message, should we aggregate loadErr got from the loop? Only the last error, which might for the lowest AppNet agent version, will be returned with the current logic.

Reply:
On a second thought, error aggregation is not required since we log the error message for each version we try load.